### PR TITLE
ENH/WIP: add a max_abs ufunc

### PR DESCRIPTION
**Pull request by ewmoore**
This is a follow on to our discussion in #5420, about adding sum_abs and max_abs ufuncs.  Testing here show this version to be almost as fast as a raw maximum.reduce for float64 and float32.  In both the SIMD loop and the BINARY_REDUCE_LOOP paths.

Still needs tests.

### CODE REVIEW

**Comment by argriffing** [comment]
Thanks for writing this!  I guess the new `PyUFunc_MaxAbsTypeResolver` could also be used for `sum_abs` type resolution later?

I see that the `USE_CHROOT=1 ARCH=i386 DIST=trusty PYTHON=3.4` Travis config and the `USE_BENTO=1` config fail the double and single precision vector norm tests, but I couldn't track down the source of the problem.

File: numpy/core/src/umath/funcs.inc.src [hunk 1]
```diff
@@ -80,6 +80,33 @@ npy_Object@Kind@(PyObject *i1, PyObject *i2)
...
 /**end repeat**/
 
+static PyObject *
+npy_ObjectMaxAbs(PyObject *i1, PyObject *i2)
+{
+    PyObject *absi1, *absi2, *result;
+    int cmp;
+
+    absi1 = PyNumber_Absolute(i1);
+    absi2 = PyNumber_Absolute(i2);
```
Comment (juliantaylor) [inline numpy/core/src/umath/funcs.inc.src]: these can probably raise exceptions and return NULL which needs handling (and a test)

File: numpy/core/src/umath/loops.c.src [hunk 2]
```diff
@@ -1617,6 +1631,42 @@ NPY_NO_EXPORT void
 }
 /**end repeat1**/
 
+NPY_NO_EXPORT void
+@TYPE@_max_abs(char **args, npy_intp *dimensions, npy_intp *steps, void* NPY_UNUSED(func))
+{
+#if @uint_matched@
+    /* Everything except signbit set, 0x7fffffffffffffff */
+    const @type@ mask0 = -0.0@c@;
```
Comment (juliantaylor) [inline numpy/core/src/umath/loops.c.src]: this trick should not be used for the generic code, as a good C citizen we cannot assume anything on the float representation. sse2 code gets away with it as its x86 only.

File: numpy/core/src/umath/loops.c.src [hunk 3]
```diff
@@ -1617,6 +1631,42 @@ NPY_NO_EXPORT void
...
+         }
+    }
+#else
+    BINARY_LOOP {
+        @type@ in1 = *(@type@ *)ip1;
+        @type@ in2 = *(@type@ *)ip2;
+        in1 = (in1 >= 0) ? in1 : -in1;
+        in2 = (in2 >= 0) ? in2 : -in2;
+        *((@type@ *)op1) = (in1 >= in2 || npy_isnan(in1)) ? in1 : in2;
+     }
```
Comment (juliantaylor) [inline numpy/core/src/umath/loops.c.src]: this ideally have a BINARY_REDUCE_LOOP specialization as it is faster for reductions

**Comment by juliantaylor** [comment]
shouldn't it be named abs_max, the absolute is applied first, similar for other functions of this type like square sum

**Comment by argriffing** [comment]
> shouldn't it be named abs_max

I'd thought that maxabs is a thing, kind of like how logsumexp is a thing.  One example is https://software.intel.com/en-us/node/502168.  But I may be wrong.

**Comment by juliantaylor** [comment]
hm maximum of absolute, sum of squares, I guess it makes sense. If other libraries use this ordering its probably best to not deviate

**Comment by juliantaylor** [comment]
instead of adding lots of new functions it would be kind of nice to have a sort of preproces argument for ufuncs that is another ufunc. Then you could e.g. do:
np.max(a, preprocess=np.abs) or possibly even recursive chains like `np.max(a, preprocess=partial(np.abs, preprocess=np.partial(np.power, 3)))`

kind of explicit unreadable lazy evaluation, on second though probably not so such a great idea

**Comment by argriffing** [comment]
> instead of adding lots of new functions it would be kind of nice to have a sort of preproces argument for ufuncs that is another ufunc.

This reminds me of the discussion near https://github.com/numpy/numpy/issues/5218#issuecomment-60137884.

**Comment by njsmith** [comment]
I think you're about 20% of the way to reinventing numexpr?

Which is not a bad thing, a version of numexpr that worked with the regular
ufunc machinery instead of reimplementing it all would be awesome.

On Tue, Jan 27, 2015 at 8:52 PM, Julian Taylor <notifications@github.com>
wrote:

> instead of adding lots of new functions it would be kind of nice to have a
> sort of preproces argument for ufuncs that is another ufunc. Then you could
> e.g. do:
> np.max(a, preprocess=np.abs) or possibly even recursive chains like np.max(a,
> preprocess=partial(np.abs, preprocess=np.partial(np.power, 3)))
>
> kind of explicit unreadable lazy evaluation, on second though probably not
> so such a great idea
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5509#issuecomment-71725432>.
>



-- 
Nathaniel J. Smith
Postdoctoral researcher - Informatics - University of Edinburgh
http://vorpus.org

**Comment by njsmith** [comment]
Basically we'd need two things:
- a way to run an arbitrary ufunc expression graph over cache-sized chunks
(not terribly hard in principle, I think -- define a ufunc loop that
recurses over some data structure calling other ufunc loops?)

- some user-friendly way to write down these expression graphs. This is
just a UI problem though and should be decoupled from the above -- instead
of awkward kwargs like prefilter= we there are lots of nice options
available e.g.

    a = delay(a)   # returns a "delay" object with trivial graph
    ma = np.maximum.reduce(np.abs(a))  # returns a "delay" object with
nontrivial graph
    ma.force()  # evaluates the graph and returns an ndarray object

__numpy_ufunc__ gets us part of the way there...

Of course one could get a minimum-viable-product by starting with
supporting relatively restricted expression graphs like pure unary nesting
and then improve from there.

On Tue, Jan 27, 2015 at 9:00 PM, Nathaniel Smith <njs@pobox.com> wrote:

> I think you're about 20% of the way to reinventing numexpr?
>
> Which is not a bad thing, a version of numexpr that worked with the
> regular ufunc machinery instead of reimplementing it all would be awesome.
>
> On Tue, Jan 27, 2015 at 8:52 PM, Julian Taylor <notifications@github.com>
> wrote:
>
>> instead of adding lots of new functions it would be kind of nice to have
>> a sort of preproces argument for ufuncs that is another ufunc. Then you
>> could e.g. do:
>> np.max(a, preprocess=np.abs) or possibly even recursive chains like np.max(a,
>> preprocess=partial(np.abs, preprocess=np.partial(np.power, 3)))
>>
>> kind of explicit unreadable lazy evaluation, on second though probably
>> not so such a great idea
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/numpy/numpy/pull/5509#issuecomment-71725432>.
>>
>
>
>
> --
> Nathaniel J. Smith
> Postdoctoral researcher - Informatics - University of Edinburgh
> http://vorpus.org
>



-- 
Nathaniel J. Smith
Postdoctoral researcher - Informatics - University of Edinburgh
http://vorpus.org

File: numpy/core/src/umath/loops.c.src (hunk 2 above)
Comment (ewmoore) [inline numpy/core/src/umath/loops.c.src]: That's a good point.

File: numpy/core/src/umath/loops.c.src (hunk 3 above)
Comment (ewmoore) [inline numpy/core/src/umath/loops.c.src]: Sure.

**Comment by ewmoore** [comment]
I've removed the code that depends on the float representation from loops.c.src.  However, This is no longer faster than just calling `abs(arr).max()` when the BINARY_LOOP path is taken. It looks to me as though gcc must generate about exactly the code I had, where I was masking out the sign bit instead of performing a comparison to find the absolute value if I use `fabs` directly.  Calling `npy_fabs` obviously doesn't cut it since gcc doesn't know anything about it.

This version is about a factor of 2 slower for the non-contiguous case.  I'm not sure what kind, if any, code changes can regain that speed in a fully portable way.  The script I've been using to benchmark things is here: https://gist.github.com/ewmoore/5f34c148b00ed461057d

Although I think allowing arbitrary combinations of ufuncs would be nice and might make a good gsoc project I'm not going to have time to do that.  So if we'd rather build that instead, lets close this.  Its also possible to only expose this function (and a hypothetical `sum_abs`) via `linalg.norm`, perhaps that represents a sort of middle road.  That being said, it seems to me that we don't want to make this slower, which this actually does right now.

**Comment by charris** [comment]
Rejigger tests by closing and opening.

**Comment by argriffing** [comment]
> Although I think allowing arbitrary combinations of ufuncs would be nice and might make a good gsoc project I'm not going to have time to do that.

I think it's reasonable to not do everything in the maximum amount of generality possible.  For example there's a short list of functions in https://software.intel.com/en-us/node/502164.  Personally I am most interested in inf-norm and 1-norm which could be implemented as `max_abs` and `sum_abs` ufuncs.

> So if we'd rather build that instead, lets close this.

In my opinion there could be value in not waiting for arbitrary combinations of ufuncs.

> Its also possible to only expose this function (and a hypothetical `sum_abs`) via linalg.norm, perhaps that represents a sort of middle road.

This seems like a good idea to me.

**Comment by charris** [comment]
@ewmoore How come?

**Comment by ewmoore** [comment]
Mostly because I don't intend to work on this further right now.  I think
its a fine idea though.

**Comment by ewmoore** [comment]
I'm very sympathetic to the idea that having all of these
dead/stalled/abandoned PRs around isn't great for the project.  Scipy has
the same problem.

**Pull request closed without merging**